File: app/src/main/java/com/supercilex/robotscouter/ui/AuthHelper.java
package com.supercilex.robotscouter.ui;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.firebase.ui.auth.AuthUI;
import com.firebase.ui.auth.ErrorCodes;
import com.firebase.ui.auth.IdpResponse;
import com.firebase.ui.auth.ResultCodes;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.firebase.appindexing.FirebaseAppIndex;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.crash.FirebaseCrash;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.data.model.User;
import com.supercilex.robotscouter.data.util.UserHelper;
import com.supercilex.robotscouter.ui.teamlist.IntentReceiver;

import static com.supercilex.robotscouter.util.AnalyticsUtilsKt.logLoginEvent;
import static com.supercilex.robotscouter.util.AnalyticsUtilsKt.updateAnalyticsUserId;
import static com.supercilex.robotscouter.util.ConstantsKt.ALL_PROVIDERS;
import static com.supercilex.robotscouter.util.ConstantsKt.FIREBASE_SCOUT_INDICES;
import static com.supercilex.robotscouter.util.ConstantsKt.FIREBASE_SCOUT_TEMPLATES;
import static com.supercilex.robotscouter.util.RemoteConfigUtilsKt.fetchAndActivate;

public final class AuthHelper implements View.OnClickListener {
    private static final int RC_SIGN_IN = 100;

    private final FragmentActivity mActivity;

    private IntentReceiver mLinkReceiver;
    private MenuItem mActionSignIn;
    private MenuItem mActionSignOut;

    public AuthHelper(FragmentActivity activity) {
        mActivity = activity;

        if (isSignedIn()) {
            initDeepLinkReceiver();
        } else {
            signInAnonymously();
        }
    }

    @Nullable
    public static FirebaseUser getUser() {
        return FirebaseAuth.getInstance().getCurrentUser();
    }

    @Nullable
    public static String getUid() {
        return getUser() == null ? null : getUser().getUid();
    }

    public static boolean isSignedIn() {
        return getUser() != null;
    }

    public static boolean isFullUser() {
        return isSignedIn() && !getUser().isAnonymous();
    }

    public static Task<FirebaseAuth> onSignedIn() {
        final TaskCompletionSource<FirebaseAuth> signInTask = new TaskCompletionSource<>();
        FirebaseAuth.getInstance().addAuthStateListener(new FirebaseAuth.AuthStateListener() {
            @Override
            public void onAuthStateChanged(@NonNull FirebaseAuth auth) {
                if (auth.getCurrentUser() == null) {
                    signInAnonymouslyDbInit();
                } else {
                    signInTask.trySetResult(auth);
                    FirebaseAuth.getInstance().removeAuthStateListener(this);
                }
            }
        });
        return signInTask.getTask();
    }

    private static Task<AuthResult> signInAnonymouslyInitBasic() {
        return FirebaseAuth.getInstance().signInAnonymously()
                .addOnSuccessListener(result -> updateAnalyticsUserId());
    }

    private static Task<AuthResult> signInAnonymouslyDbInit() {
        return signInAnonymouslyInitBasic().addOnSuccessListener(result -> DatabaseInitializer.init());
    }

    public void initMenu(Menu menu) {
        mActionSignIn = menu.findItem(R.id.action_sign_in);
        mActionSignOut = menu.findItem(R.id.action_sign_out);
        toggleMenuSignIn(isSignedIn() && !getUser().isAnonymous());
    }

    private void initDeepLinkReceiver() {
        if (mLinkReceiver == null) {
            mLinkReceiver = IntentReceiver.init(mActivity);
        }
    }

    public void signIn() {
        mActivity.startActivityForResult(
                AuthUI.getInstance().createSignInIntentBuilder()
                        .setAvailableProviders(ALL_PROVIDERS)
                        .setLogo(R.drawable.ic_logo)
                        .setTosUrl("https://supercilex.github.io/privacy-policy/")
                        .setIsAccountLinkingEnabled(true)
                        .build(),
                RC_SIGN_IN);
    }

    private void signInAnonymously() {
        signInAnonymouslyDbInit()
                .addOnSuccessListener(mActivity, result -> initDeepLinkReceiver())
                .addOnFailureListener(mActivity, e ->
                        Snackbar.make(mActivity.findViewById(R.id.root),
                                      R.string.anonymous_sign_in_failed,
                                      Snackbar.LENGTH_LONG)
                                .setAction(R.string.sign_in, this)
                                .show());
    }

    public void signOut() {
        AuthUI.getInstance()
                .signOut(mActivity)
                .addOnSuccessListener(aVoid -> {
                    signInAnonymouslyInitBasic();
                    FirebaseAppIndex.getInstance().removeAll();
                })
                .addOnSuccessListener(mActivity, aVoid -> toggleMenuSignIn(false));
    }

    public void showSignInResolution() {
        Snackbar.make(mActivity.findViewById(R.id.root),
                      R.string.sign_in_required,
                      Snackbar.LENGTH_LONG)
                .setAction(R.string.sign_in, this)
                .show();
    }

    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == RC_SIGN_IN) {
            IdpResponse response = IdpResponse.fromResultIntent(data);

            if (resultCode == ResultCodes.OK) {
                Snackbar.make(mActivity.findViewById(R.id.root),
                              R.string.signed_in,
                              Snackbar.LENGTH_LONG)
                        .show();
                toggleMenuSignIn(true);

                initDeepLinkReceiver();

                UserHelper userHelper = new User.Builder(getUid())
                        .setEmail(getUser().getEmail())
                        .setName(getUser().getDisplayName())
                        .setPhotoUrl(getUser().getPhotoUrl())
                        .build()
                        .getHelper();
                userHelper.add();
                if (response != null) userHelper.transferData(response.getPrevUid());

                logLoginEvent();

                return true;
            } else {
                if (response == null) return false; // User cancelled sign in

                if (response.getErrorCode() == ErrorCodes.NO_NETWORK) {
                    Snackbar.make(mActivity.findViewById(R.id.root),
                                  R.string.no_connection,
                                  Snackbar.LENGTH_LONG)
                            .setAction(R.string.try_again, this)
                            .show();
                    return false;
                }

                Snackbar.make(mActivity.findViewById(R.id.root),
                              R.string.sign_in_failed,
                              Snackbar.LENGTH_LONG)
                        .setAction(R.string.try_again, this)
                        .show();
            }
        }

        return false;
    }

    private void toggleMenuSignIn(boolean isSignedIn) {
        mActionSignIn.setVisible(!isSignedIn);
        mActionSignOut.setVisible(isSignedIn);
    }

    @Override
    public void onClick(View v) {
        signIn();
    }

    private static final class DatabaseInitializer implements ValueEventListener, OnSuccessListener<Void> {
        private static final String SHOULD_CACHE_DB = "should_cache_db";

        private DatabaseInitializer() {
            fetchAndActivate().addOnSuccessListener(this);
            FIREBASE_SCOUT_INDICES.addListenerForSingleValueEvent(this);
        }

        public static void init() {
            new DatabaseInitializer();
        }

        @Override
        public void onSuccess(Void aVoid) {
            if (FirebaseRemoteConfig.getInstance().getBoolean(SHOULD_CACHE_DB)) {
                FIREBASE_SCOUT_TEMPLATES.addListenerForSingleValueEvent(this);
            }
        }

        @Override
        public void onDataChange(DataSnapshot snapshot) {
            // This allows the database to work offline without any setup
        }

        @Override
        public void onCancelled(DatabaseError error) {
            FirebaseCrash.report(error.toException());
        }
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/AuthHelper.kt
package com.supercilex.robotscouter.ui

import android.content.Intent
import android.support.design.widget.Snackbar
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.firebase.ui.auth.AuthUI
import com.firebase.ui.auth.ErrorCodes
import com.firebase.ui.auth.IdpResponse
import com.firebase.ui.auth.ResultCodes
import com.google.firebase.appindexing.FirebaseAppIndex
import com.google.firebase.auth.FirebaseUser
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.data.model.User
import com.supercilex.robotscouter.data.model.helper
import com.supercilex.robotscouter.ui.teamlist.IntentForwarder
import com.supercilex.robotscouter.ui.teamlist.TeamListActivity
import com.supercilex.robotscouter.util.ALL_PROVIDERS
import com.supercilex.robotscouter.util.getUid
import com.supercilex.robotscouter.util.getUser
import com.supercilex.robotscouter.util.isFullUser
import com.supercilex.robotscouter.util.isSignedIn
import com.supercilex.robotscouter.util.logLoginEvent
import com.supercilex.robotscouter.util.signInAnonymouslyDbInit
import com.supercilex.robotscouter.util.signInAnonymouslyInitBasic

class AuthHelper(private val activity: TeamListActivity) : View.OnClickListener {
    private val rootView = activity.findViewById(R.id.root)

    private val linkReceiver by lazy { IntentForwarder(activity) }
    private var actionSignIn: MenuItem? = null
    private var actionSignOut: MenuItem? = null

    init {
        if (isSignedIn()) initDeepLinkReceiver()
        else signInAnonymously()
    }

    fun initMenu(menu: Menu) {
        actionSignIn = menu.findItem(R.id.action_sign_in)
        actionSignOut = menu.findItem(R.id.action_sign_out)
        toggleMenuSignIn(isFullUser())
    }

    private fun initDeepLinkReceiver() = linkReceiver

    fun signIn() = activity.startActivityForResult(
            AuthUI.getInstance().createSignInIntentBuilder()
                    .setAvailableProviders(ALL_PROVIDERS)
                    .setLogo(R.drawable.ic_logo)
                    .setPrivacyPolicyUrl("https://supercilex.github.io/Robot-Scouter/privacy-policy/")
                    .setIsAccountLinkingEnabled(true)
                    .build(),
            RC_SIGN_IN)

    private fun signInAnonymously() = signInAnonymouslyDbInit()
            .addOnSuccessListener(activity) { initDeepLinkReceiver() }
            .addOnFailureListener(activity) {
                Snackbar.make(rootView, R.string.anonymous_sign_in_failed, Snackbar.LENGTH_LONG)
                        .setAction(R.string.sign_in, this)
                        .show()
            }

    fun signOut() = AuthUI.getInstance()
            .signOut(activity)
            .addOnSuccessListener {
                signInAnonymouslyInitBasic()
                FirebaseAppIndex.getInstance().removeAll()
            }
            .addOnSuccessListener(activity) { toggleMenuSignIn(false) }

    fun showSignInResolution() =
            Snackbar.make(rootView, R.string.sign_in_required, Snackbar.LENGTH_LONG)
                    .setAction(R.string.sign_in, this)
                    .show()

    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode == RC_SIGN_IN) {
            val response: IdpResponse? = IdpResponse.fromResultIntent(data)

            if (resultCode == ResultCodes.OK) {
                Snackbar.make(rootView,
                        R.string.signed_in,
                        Snackbar.LENGTH_LONG)
                        .show()
                toggleMenuSignIn(true)

                initDeepLinkReceiver()

                val user: FirebaseUser = getUser()!!
                val userHelper =
                        User(getUid()!!, user.email, user.displayName, user.photoUrl).helper
                userHelper.add()
                if (response != null) userHelper.transferData(response.prevUid)

                logLoginEvent()

                return true
            } else {
                if (response == null) return false // User cancelled sign in

                if (response.errorCode == ErrorCodes.NO_NETWORK) {
                    Snackbar.make(rootView, R.string.no_connection, Snackbar.LENGTH_LONG)
                            .setAction(R.string.try_again, this)
                            .show()
                    return false
                }

                Snackbar.make(rootView, R.string.sign_in_failed, Snackbar.LENGTH_LONG)
                        .setAction(R.string.try_again, this)
                        .show()
            }
        }

        return false
    }

    private fun toggleMenuSignIn(isSignedIn: Boolean) {
        actionSignIn?.isVisible = !isSignedIn
        actionSignOut?.isVisible = isSignedIn
    }

    override fun onClick(v: View) = signIn()

    companion object {
        private val RC_SIGN_IN = 100
    }
}
